### PR TITLE
plumbing to support refraction

### PR DESCRIPTION
**Pull request by pixelflinger**

### CODE REVIEW

File: libs/filabridge/include/filament/MaterialEnums.h [hunk 1]
```diff
@@ -152,6 +152,36 @@ enum MaterialDomain : uint8_t {
...
+    SUBSURFACE_COLOR,        // float3, subsurface and cloth shading models only
+    SHEEN_COLOR,             // float3, cloth shading model only
+    SPECULAR_COLOR,          // float3, specular-glossiness shading model only
+    GLOSSINESS,              // float,  specular-glossiness shading model only
+    EMISSIVE,                // float4, all shading models
+    NORMAL,                  // float3, all shading models only, except unlit
+    POST_LIGHTING_COLOR,     // float4, all shading models
+    CLIP_SPACE_TRANSFORM,    // mat4,   vertex shader only
+    IOR,                     // float,  index of refraction (activate refraction)
+    ABSORPTION,              // float,  how much light is absorbed by the material
```
Comment (romainguy) [inline libs/filabridge/include/filament/MaterialEnums.h]: This needs to be a float3, not a float. Absorption is wavelength dependent.

File: libs/filabridge/include/filament/MaterialEnums.h [hunk 2]
```diff
@@ -152,6 +152,36 @@ enum MaterialDomain : uint8_t {
...
+    TRANSMISSION,            // float,  how much light is refracted through the material
+    THIN_LAYER_THICKNESS,    // float,  thin layer thickness
```
Comment (romainguy) [inline libs/filabridge/include/filament/MaterialEnums.h]: Why not reuse thickness? Also we probably don't want to have the concept of thickness for thin-walled transparency. I would assume that setting the thickness would trigger the change from thin transparency to volume transparency.

File: libs/filamat/tests/test_filamat.cpp [hunk 3]
```diff
@@ -414,6 +414,30 @@ TEST_F(MaterialCompiler, StaticCodeAnalyzerClearCoat) {
...
 }
 
+TEST_F(MaterialCompiler, StaticCodeAnalyzerIor) {
+    std::string fragmentCode(R"(
+        void material(inout MaterialInputs material) {
+            prepareMaterial(material);
+            material.ior = 1.5;
+            material.absorption = 0.0;
+            material.transmission = 0.96;
+            material.thin_layer_thickness = 0.01;
```
Comment (romainguy) [inline libs/filamat/tests/test_filamat.cpp]: We don't use `_` in property names.

File: shaders/src/material_inputs.fs [hunk 4]
```diff
@@ -57,6 +57,13 @@ struct MaterialInputs {
 #if defined(MATERIAL_HAS_POST_LIGHTING_COLOR)
     vec4  postLightingColor;
 #endif
+
+#if defined(MATERIAL_HAS_IOR)
+    float ior;
+    float absorption;
```
Comment (romainguy) [inline shaders/src/material_inputs.fs]: Not sure we want to enable all properties when ior is set. We should discuss this.

File: libs/filabridge/include/filament/MaterialChunkType.h [hunk 5]
```diff
@@ -45,6 +45,7 @@ enum UTILS_PUBLIC ChunkType : uint64_t {
     MaterialMetal = charTo64bitNum("MAT_METL"),
     MaterialShaderModels = charTo64bitNum("MAT_SMDL"),
     MaterialSamplerBindings = charTo64bitNum("MAT_SAMP"),   // no longer used
+    MaterialProperties = charTo64bitNum("MAT_PROP"),
```
Comment (bejado) [inline libs/filabridge/include/filament/MaterialChunkType.h]: We should probably deprecate the `MaterialSpecularAntiAliasing` chunk and instead use this new one to determine if the material has specular anti-aliasing.

File: libs/filamat/src/Enums.cpp [hunk 6]
```diff
@@ -41,6 +41,10 @@ std::unordered_map<std::string, Property> Enums::mStringToProperty = {
         { "normal",              Property::NORMAL },
         { "postLightingColor",   Property::POST_LIGHTING_COLOR },
         { "clipSpaceTransform",  Property::CLIP_SPACE_TRANSFORM },
+        { "ior",                 Property::IOR },
+        { "absorption",          Property::ABSORPTION },
+        { "transmission",        Property::TRANSMISSION },
+        { "thin_layer_thickness",Property::THIN_LAYER_THICKNESS },
```
Comment (bejado) [inline libs/filamat/src/Enums.cpp]: `thinLayerThickness` for consistency?

File: shaders/src/material_inputs.fs [hunk 7]
```diff
@@ -57,6 +57,13 @@ struct MaterialInputs {
 #if defined(MATERIAL_HAS_POST_LIGHTING_COLOR)
     vec4  postLightingColor;
 #endif
+
+#if defined(MATERIAL_HAS_IOR)
+    float ior;
```
Comment (bejado) [inline shaders/src/material_inputs.fs]: We should give these default values inside of `initMaterial`.

File: libs/filabridge/include/filament/MaterialChunkType.h (hunk 5 above)
Comment (romainguy) [reply libs/filabridge/include/filament/MaterialChunkType.h]: No this is just for properties set inside the shader. specularAA is different.

Comment (bejado) [reply libs/filabridge/include/filament/MaterialChunkType.h]: Ah I see

**Pull request merged**